### PR TITLE
Reduced the size of the hazard curves

### PR DESCRIPTION
**Pull request by micheles**
This requires https://github.com/gem/oq-hazardlib/pull/453/.
By replacing the hazard curves with ProbabilityMaps the data transfer can be reduced significantly.
Here is the case for a reduced SHARE calculation (with less sources to emphasize the speedup) on Wilson:
```
   current master
   ============================== ======== ========= ======
   operation                      time_sec memory_mb counts
   ============================== ======== ========= ======
   total classical                371      96        50    
   making contexts                126      0.0       27,574
   computing poes                 109      0.0       52,444
   total runtime                  40       1,035     1     
   aggregate curves               2.735    33        50    
   compute and save statistics    0.352    0.0       1     
   combine curves_by_rlz          0.020    0.0       1     
   save curves_by_trt_gsim        0.017    0.020     1     
   ============================== ======== ========= ======
   classical_tot_received          989,730,489

   this branch
   ============================== ======== ========= ======
   operation                      time_sec memory_mb counts
   ============================== ======== ========= ======
   total classical                259      15        50    
   making contexts                109      0.0       27,574
   computing poes                 108      0.0       13,111
   total runtime                  39       98        1     
   aggregate curves               0.385    0.238     50    
   compute and save statistics    0.374    28        1     
   combine curves_by_rlz          0.148    0.0       1     
   save curves_by_trt_id          0.137    11        1     
   ============================== ======== ========= ======
   classical_tot_received          17,046,411  (1/58)
```
The data transfer has been reduced by a factor of 58 (!) The memory occupation has been reduced
from 96 MB to 15 MB. The total runtime is still more or less the same, because it is dominated by the slowest task. Another analysis for a reduced SARA model gives similar results: memory occupation reduced by 3 times, data transfer reduced by 37 times, total runtime reduced by only 10%.  

The tests are running here: https://ci.openquake.org/job/zdevel_oq-risklib/1605

### CODE REVIEW

File: openquake/calculators/classical.py [hunk 1]
```diff
@@ -307,38 +278,34 @@ def store_source_info(self, curves_by_trt_gsim):
...
 
         with self.monitor('combine curves_by_rlz', autoflush=True):
-            zc = zero_curves(len(self.sitecol.complete), self.oqparam.imtls)
-            curves_by_rlz = self.rlzs_assoc.combine_curves(
-                curves_by_trt_gsim, agg_curves, zc)
-        self.save_curves(curves_by_rlz)
+            curves_by_rlz = self.rlzs_assoc.combine_curves(curves_by_trt_gsim)
+
+        self.save_curves({rlz: array_of_curves(curves, nsites, imtls)
+                          for rlz, curves in curves_by_rlz.items()})
```
Comment (ptormene) [inline openquake/calculators/classical.py]: Please change to pmaps_by_rlz

File: openquake/commonlib/oqvalidation.py [hunk 2]
```diff
@@ -285,7 +284,8 @@ def no_imls(self):
         """
         Return True if there are no intensity measure levels
         """
-        return all(ls is None for ls in self.imtls.values())
+        return all(ls is None or numpy.isnan(ls).any()
```
Comment (ptormene) [inline openquake/commonlib/oqvalidation.py]: Please check if this is needed

**Comment by ptormene** [comment]
Besides minor comments, LGTM

File: openquake/calculators/classical.py (hunk 1 above)
Comment (micheles) [inline openquake/calculators/classical.py]: On second thought, I will keep the name as it is for the moment

File: openquake/commonlib/oqvalidation.py (hunk 2 above)
Comment (micheles) [inline openquake/commonlib/oqvalidation.py]: It is not needed anymore

**Pull request merged**